File: frontend/lib/json.dart
class Message {
  MessageType type;
  Map<String, dynamic> data;  

  Message(this.type, this.data);

  Message.fromJson(Map<String, dynamic> json) :
    type = MessageType.values[json['type']],
    data = json['data'];

  Map<String, dynamic> toJson() => {
    'type': type.toString(),
    'data': data
  };

  Message.joinGame(int gameId) :
    type = MessageType.JoinGame,
    data = {
      'game': gameId
    };

  Message.setName(String name) :
    type = MessageType.SetName,
    data = {
      'name': name
    };

  Message.startGame() :
    type = MessageType.StartGame,
    data = {};

  Message.createGame() :
    type = MessageType.CreateGame,
    data = {};
}

enum MessageType {
  Info,
  Error,
  JoinGame,
  SetName,
  StartGame,
  CreateGame
}

File: backend/src/json.rs
use serde::{Deserialize, Serialize};

impl Message {
    pub fn to_string(&self) -> String {
        serde_json::to_string(self).unwrap()
    }
}


#[derive(Serialize, Deserialize)]
pub enum Message {
    Info { message: String },
    Error { message: String },
    JoinGame { game: u32 },
    SetName { name: String },
    StartGame,
    CreateGame,
}